addons/hr_expense/models/hr_expense.py

```diff
@@ -312,3 +286,130 @@ def action_get_attachment_view(self):
+
+    _name = "hr.expense.sheet"
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+    _description = "Expense Sheet"
+    _order = "date desc"
+
+    name = fields.Char(string='Expense Sheet Description', readonly=True, required=True, states={'draft': [('readonly', False)]})
+    date = fields.Date(readonly=True, states={'draft': [('readonly', False)]}, default=fields.Date.context_today, string="Date")
+    expense_line_ids = fields.Many2many('hr.expense', string='Expense Lines', states={'done': [('readonly', True)], 'post': [('readonly', True)]})
+    state = fields.Selection([('draft', 'To Submit'),
+                              ('submit', 'Submitted'),
+                              ('approve', 'Approved'),
+                              ('post', 'Waiting Payment'),
+                              ('done', 'Paid'),
+                              ('cancel', 'Refused')
+                              ], string='Status', index=True, readonly=True, track_visibility='onchange', copy=False, default='draft', required=True,
+        help='When the expense sheet request is created the status is \'To Submit\'.\n It is submitted by the employee and request is sent to manager, the status is \'Submitted\'.\
+        \nIf the manager approve it, the status is \'Approved\'.\n If the accountant genrate the accounting entries for the expense sheet request, the status is \'Waiting Payment\'.')
+    employee_id = fields.Many2one('hr.employee', string="Employee", required=True, readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1))
+    payment_status = fields.Boolean(string='Payment received by employee', help='The employee greens this button when the payment has been received.')
+    user_valid = fields.Many2one('res.users', 'Validation By', readonly=True, copy=False, states={'draft':[('readonly',False)], 'submit':[('readonly',False)]})
+    total_amount = fields.Float(string='Total Amount', store=True, compute='_compute_amount', digits=dp.get_precision('Account'))
+    company_id = fields.Many2one('res.company', string='Company', readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env.user.company_id)
+    currency_id = fields.Many2one('res.currency', string='Currency', readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env.user.company_id.currency_id)
+    attachment_number = fields.Integer(compute='_compute_attachment_number', string='Number of Attachments')
+
+    @api.multi
+    def write(self, vals):
+        old_line_ids = set(self.expense_line_ids.ids)
+        res = super(HrExpenseSheet, self).write(vals)
+        if vals.get('expense_line_ids'):
+            new_line_ids = set(self.expense_line_ids.ids)
+            rec_to_update = list(old_line_ids - new_line_ids)
+            self.env['hr.expense'].browse(rec_to_update).write({'sheet_id': None, 'state': 'draft'})
```

Comment:
Moreover, I think it should call a reset() method on the expense lines that do the reset, instead of doing it here. Let us delegate the lines logic to the lines model.